### PR TITLE
Add scenario json schema

### LINKED ISSUE #346: Export JSON schema for scenarios
It would be nice to get editor support for writing scenarios.

[JSON schemas for YAML](https://json-schema.org/learn/getting-started-step-by-step.html) seem to do the job, but can we generate them somehow without much work?

### PR DESCRIPTION
**Pull request by xsebek**
- added JSON schemas for scenarios, entities and recipes
- closes #346

### CODE REVIEW

**Comment by xsebek** [comment]
To test this in VS Code, add this to `setting.json`:
```json
"yaml.schemas": {
    "https://raw.githubusercontent.com/swarm-game/swarm/13da5d6cdc0308e3febaedf2350264761e53cc70/data/scenarios/schema.json": [
        "data/scenarios/*.yaml",
        "data/scenarios/**/*.yaml"
    ]
}
```
It seems to work pretty well, but I did not copy the description of inner objects yet:
image

**Comment by xsebek** [comment]
To validate from command line, run:
```bash
# create a typo
sed -i 's/world:/worlds:/g' scenarios/creative.yaml
# validate
yq eval scenarios/creative.yaml -o json | jsonschema scenarios/schema.json
# {'$schema': 'https://raw.githubusercontent.com/swarm-game/swarm/13da5d6cdc0308e3febaedf2350264761e53cc70/data/scenarios/schema.json', 'name': 'Creative game', 'description': "No constraints!  Do whatever you want, create any items you want out of thin air, build robots to your heart's content.", 'creative': True, 'robots': [{'name': 'base', 'loc': [0, 0], 'dir': [0, 1], 'heavy': True, 'display': {'char': 'Ω', 'attr': 'robot'}}], 'worlds': {'seed': None, 'offset': True}}:
# 'world' is a required property
```

File: data/schema/display.json [hunk 1]
```diff
@@ -0,0 +1,62 @@
+{
+    "$schema": "http://json-schema.org/draft-07/schema#",
+    "$id": "https://raw.githubusercontent.com/swarm-game/swarm/scheme/data/schema/display.json",
```
Comment (xsebek) [inline data/schema/display.json:3]: Before merging, this needs to be changed to point to the `main` branch:
```bash
sed 's|swarm-game/swarm/scheme|warm-game/swarm/main|g
```

**Comment by xsebek** [comment]
@byorgey could you please give this a try in Emacs? There is supposed to be some LSP support for it:
- https://emacs-lsp.github.io/lsp-mode/page/lsp-yaml/

Also, I am thinking about adding a small GitHub action, that would check all scenarios against the schema. :thinking: 
The same could be done for `entities.yaml` and `recipes.yaml`, once I change their schemas to arrays.

**Comment by byorgey** [comment]
@xsebek you mention `scenarios/schema.json` several times in your comments above, but there is no such file.  Did you mean `data/schema/scenario.json`?

**Comment by byorgey** [comment]
Also, I can't seem to get it to validate on the command line.  `jsonschema` seems to show no output no matter what I do.

**Comment by byorgey** [comment]
I got the `lsp-yaml` package installed in emacs, it seems to work.  However, I am really not sure how to point it at the schemas.  Emacs does not seem to have an equivalent of `setting.json` and I am not sure how to configure the `lsp-yaml` server.

**Review by byorgey — APPROVED** [review]

**Comment by xsebek** [comment]
@byorgey re: `data/schema`, yes, I moved the files over to a new directory, as I plan to use it also for other YAML files.

It could be it shows no output because the file has no errors, did you try the `sed` command? Also, you could have an old version of `jsonschema`, I installed mine with `pip`.

I updated the bash commands above, so they should work now. :wink: 

---

If you are using `flycheck`, then it should in turn use `js-yaml` which can be configured to check schemas somehow. :confused:
https://www.flycheck.org/en/latest/languages.html#syntax-checker-yaml-jsyaml

**Comment by byorgey** [comment]
What version of `jsonschema` do you have? I have 3.2.0.  When I `pip install jsonschema` it says it is already installed.  And yes, I definitely used `sed` to introduce an error.  Even when I do `cat | jsonschema data/schema/scenario.json` and type in nonsense, it does not produce any output.

**Comment by xsebek** [comment]
Well, [that version is from 2019](https://pypi.org/project/jsonschema/3.2.0/) and it may not take input from stdin. :thinking:  You can try the `--input` flag.

I have a more recent one:
```bash
pip --version 
# pip 22.1 from /home/sebeko/.local/lib/python3.10/site-packages/pip (python 3.10)
jsonschema --version
# 4.9.1
echo {} | jsonschema data/schema/scenario.json
# {}: 'name' is a required property
# {}: 'world' is a required property
# {}: 'robots' is a required property
```

**Pull request merged**